### PR TITLE
[SPARK-32310][ML][PySpark][3.0] ML params default value parity

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
backporting the changes to 3.0
set params default values in trait Params for feature and tuning in both Scala and Python.

### Why are the changes needed?
Make ML has the same default param values between estimator and its corresponding transformer, and also between Scala and Python.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing and modified tests

### CODE REVIEW

**Comment by huaxingao** [comment]
cc @srowen @viirya @zhengruifeng

**Comment by zhengruifeng** [comment]
retest this please

**Comment by srowen** [comment]
Jenkins retest this please

**Comment by huaxingao** [comment]
retest this please

**Comment by huaxingao** [comment]
retest this please

**Comment by huaxingao** [comment]
retest this please

**Comment by viirya** [comment]
retest this please

**Comment by viirya** [comment]
For backport PR, can you add [3.0] to the PR title?

**Comment by viirya** [comment]
Is this just a backport of #29112? Looks it it is bigger change than #29112?

**Comment by huaxingao** [comment]
@viirya It's backporting #29122 and #29153. I haven't finished #29153 yet. I want to finish 3.0 first because seems the plan is to cut 3.0.1 pretty soon.

**Comment by Fokko** [comment]
#29122 Isn't included in the PR AFAIK.

**Comment by huaxingao** [comment]
Sorry, fixed. @Fokko

File: mllib/src/test/scala/org/apache/spark/ml/util/DefaultReadWriteTest.scala [hunk 1]
```diff
@@ -64,7 +64,12 @@ trait DefaultReadWriteTest extends TempDirectory { self: Suite =>
             case (Array(values), Array(newValues)) =>
               assert(values === newValues, s"Values do not match on param ${p.name}.")
             case (value, newValue) =>
-              assert(value === newValue, s"Values do not match on param ${p.name}.")
+              if (value.isInstanceOf[Double] && value.asInstanceOf[Double].isNaN) {
```
Comment (zhengruifeng) [inline mllib/src/test/scala/org/apache/spark/ml/util/DefaultReadWriteTest.scala:67]: Why we need this check?

If necessary, may change to:
```
case (value: Double, newValue: Double) =>
      assert(value.isNaN && newValue.isNaN || value == newValue, ...)
```

Comment (huaxingao) [reply mllib/src/test/scala/org/apache/spark/ml/util/DefaultReadWriteTest.scala]: Before my change, this ```setDefault(strategy -> Imputer.mean, missingValue -> Double.NaN)``` is in Imputer. Now I moved this to ImputerParams so ImputerModel also has the default value of ```missingValue``` set to ```Double.NaN```. In ```DefaultReadWriteTest```, it compares the default value of ```missingValue``` in the original model and the loaded model, so I need to make the NaN comparison work.

**Comment by huaxingao** [comment]
retest this please

**Comment by huaxingao** [comment]
retest this please

**Comment by huaxingao** [comment]
retest this please

**Comment by huaxingao** [comment]
retest this please

**Comment by huaxingao** [comment]
retest this please

**Comment by huaxingao** [comment]
retest this please

File: python/pyspark/ml/tuning.py [hunk 2]
```diff
@@ -545,6 +551,10 @@ class _TrainValidationSplitParams(_ValidatorParams):
     trainRatio = Param(Params._dummy(), "trainRatio", "Param for ratio between train and\
      validation data. Must be between 0 and 1.", typeConverter=TypeConverters.toFloat)
 
+    def __init__(self, *args):
```
Comment (viirya) [inline python/pyspark/ml/tuning.py:554]: You added `*args` many places. Just to make sure it is for consistency?

**Comment by viirya** [comment]
retest this please

**Review by viirya — COMMENTED** [review]
Looks good overall.

File: python/pyspark/ml/tuning.py (hunk 2 above)
Comment (huaxingao) [reply python/pyspark/ml/tuning.py:554]: Yes, it's for consistency.

**Comment by viirya** [comment]
retest this please

**Comment by huaxingao** [comment]
Jenkins, please make pip packaging tests pass! retest this please

**Review by zhengruifeng — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by huaxingao** [comment]
@srowen Are you Okay with this?

**Comment by srowen** [comment]
Same as the version for master - looks OK if there are no API or behavior changes (that are not bug fixes).

**Comment by huaxingao** [comment]
Merged to 3.0. Thank you all for reviewing!

**Pull request closed without merging**